src/main/java/com/statsig/androidsdk/Statsig.kt

```diff
@@ -23,431 +33,423 @@ interface IStatsigCallback {
+            application.registerActivityLifecycleCallbacks(lifecycleListener)
+            val sharedPrefs: SharedPreferences = getSharedPrefs()
+            logger = StatsigLogger(
+                sdkKey,
+                options.api,
+                statsigMetadata,
+                sharedPrefs,
+                statsigNetwork
+            )
             loadFromCache()
 
-            val initResponse = StatsigNetwork.initialize(
-                this.options.api,
+            val initResponse = statsigNetwork.initialize(
+                options.api,
                 sdkKey,
                 user,
-                this.statsigMetadata,
-                this.options.initTimeoutMs,
+                statsigMetadata,
+                options.initTimeoutMs,
             )
 
-            val instance = this
-            runBlocking {
-                instance.setState(initResponse)
-            }
+            setState(initResponse)
 
-            if (this.options.enableAutoValueUpdate) {
-                instance.pollForUpdates()
+            if (options.enableAutoValueUpdate) {
+                pollForUpdates()
             }
 
-            if (sharedPrefs != null) {
-                StatsigNetwork.apiRetryFailedLogs(
-                    this.options.api,
-                    sdkKey,
-                    sharedPrefs
-                )
-            }
+            statsigNetwork.apiRetryFailedLogs(options.api, sdkKey, sharedPrefs)
         }
+    }
 
-        /**
-         * Check the value of a Feature Gate configured in the Statsig console for the initialized
-         * user
-         * @param gateName the name of the feature gate to check
-         * @return the value of the gate for the initialized user, or false if not found
-         * @throws IllegalStateException if the SDK has not been initialized
-         */
-        @JvmStatic
-        fun checkGate(gateName: String): Boolean {
-            enforceInitialized("checkGate")
-            var res = APIFeatureGate(gateName)
-            if (this.state != null) {
-                res = this.state!!.checkGate(StatsigUtil.getHashedString(gateName))
-            }
-            this.logger.logGateExposure(gateName, res.value, res.ruleID, this.user)
-            return res.value
+    /**
+     * Check the value of a Feature Gate configured in the Statsig console for the initialized
+     * user
+     * @param gateName the name of the feature gate to check
+     * @return the value of the gate for the initialized user, or false if not found
+     * @throws IllegalStateException if the SDK has not been initialized
+     */
+    @JvmStatic
+    fun checkGate(gateName: String): Boolean {
+        enforceInitialized("checkGate")
+        val res =
+            state?.checkGate(StatsigUtil.getHashedString(gateName)) ?: APIFeatureGate(gateName)
+        statsigScope.launch {
+            logger.logGateExposure(gateName, res.value, res.ruleID, user)
         }
```

Comment:
Yes, this is great.  Regarding the similar change on the server sdk - the main breaking change was making the logevent api asynchronous.  I think we could do something similar there to continue to support a synchronous logEvent api.

Reply:
So the calls are "asynchronous", but sort of... Everything runs on the main thread except for network calls, which makes the calls run practically immediately, unless the main thread is under stress, which is actually a huge problem as it would affect UI. The concern is that if the logger becomes synchronous, it can cause UI performance problem when this call is called on the main thread, as it would then block the thread, instead of suspending the execution of code.

One thing I would suggest is writing all logs to a file (appending a file is quick and easy), and flush it to network at an interval. This would provide a quick caching mechanism that would persist even if the app process is killed (which happens _often_).

Reply:
We already flush periodically anyway, so its just a matter of swapping out tracking logs in the state of the logger vs writing to a file.

I suppose the suggestion is to post to the IO thread for that, rather than doing the work on the main thread?  If so, that makes perfect sense.